### ISSUE TITLE
Converted OBB ONNX model output is different with my custom code

### ISSUE THREAD

**Issue by dilwolf** [ultralytics/ultralytics#21422]
Hello there,

I have successfully fine-tuned the YOLOv11s OBB model on a single-person class and converted it to ONNX using ultralytics code like below:

```
from ultralytics import YOLO
model = YOLO("person-obb.pt")
model.export(format='onnx', simplify=True, nms=True)
```

For later implementation in C++, I wrote a Python inference script for my ONNX model like below:

```
 def preprocess_image(image_path, imgsz):
    image = cv2.imread(image_path)
    if image is None:
        raise ValueError(f"Failed to load image at {image_path}") 
    orig_image = image.copy()      
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)  # BGR to RGB, possibly using letterbox 
    orig_shape = image.shape[:2]  # H, W 
    image = cv2.resize(image, (imgsz, imgsz))  # Resize to model input size (imgsz x imgsz)
    image = image.astype(np.float32) / 255.0  # Normalize to [0, 1]
    image = np.transpose(image, (2, 0, 1)) # HWC -> CHW ==> C x H x W
    return orig_image, image, orig_shape

def xywhr2xyxyxyxy(x):
    x = np.asarray(x)
    ctr = x[..., :2]  # center coordinates
    w, h, angle = (x[..., i:i+1] for i in range(2, 5))
    cos_value = np.cos(angle)
    sin_value = np.sin(angle)
    # Vector components
    vec1 = np.concatenate([w / 2 * cos_value, w / 2 * sin_value], axis=-1)
    vec2 = np.concatenate([-h / 2 * sin_value, h / 2 * cos_value], axis=-1)
    pt1 = ctr + vec1 + vec2
    pt2 = ctr + vec1 - vec2
    pt3 = ctr - vec1 - vec2
    pt4 = ctr - vec1 + vec2
    return np.stack([pt1, pt2, pt3, pt4], axis=-2)

def postprocess_obb(prediction, orig_shape, conf_thresh=0.25, max_det=300): 
    output = [pred[pred[:, 4] > conf_thresh][:max_det] for pred in prediction]  
    if isinstance(output, (list, tuple)):
        output = output[0]  # select only inference output
    boxes = np.concatenate([output[:, :4], output[:, -1:]], axis=1)
    scores = output[:, 4]
    class_ids = output[:, 5]
    # Scale bboxes
    scale_x, scale_y = orig_shape[1] / imgsz, orig_shape[0] / imgsz
    boxes[:, 0] *= scale_x  # cx
    boxes[:, 1] *= scale_y  # cy
    boxes[:, 2] *= scale_x  # w
    boxes[:, 3] *= scale_y  # h 
    return boxes, scores, class_ids
        
def plot_bboxes(original_frame, bboxes, labels, conf_scores, color):
    labels = [int(l) for l in labels]
    bboxes = xywhr2xyxyxyxy(bboxes)
    for bbox in bboxes: # loop over all bboxes     
        cv2.polylines(original_frame, [np.asarray(bbox, dtype=int)], True, color, 2) # draw OBB bbox
    return original_frame

onnx_model = onnxruntime.InferenceSession("person-obb.onnx")
image_path = "Capture2.png"
imgsz = 1024
color=(0, 255, 0)

orig_image, processed_frame, orig_shape = preprocess_image(image_path, imgsz)
output = onnx_model.run(None, {onnx_model.get_inputs()[0].name: np.expand_dims(processed_frame, axis=0)})
boxes, scores, class_ids = postprocess_obb(output[0], orig_shape, conf_thresh=0.25)
ann_img = plot_bboxes(orig_image, boxes, class_ids, scores, color)
# Save output
base_name = "onnx_result_png.png"
cv2.imwrite(base_name, ann_img)
```

The problem here is that my code results and the results of the Ultralytics inference code are a little bit different on the same input image and the same ONNX model. The results are the same with some input images, but with some of the same input images, the results are different. Below is ultralytics code:

```
from ultralytics import YOLO
model = YOLO("person-obb.onnx", task="obb")
results = model("Capture2.png", imgsz=1024, save=True)
```

I am not sure what I am doing wrong with my custom Python code?? I suppose the problem is in the preprocessing step (I could not find this step's code from the ultralytics package) because the `bboxes`  are different  when I compare `bboxes` after confidence thresholding between my custom code and the ultralytics code. I am asking this because it is not expected, right?

FIY: I am sharing a snapshot of the ONNX model in Netron.

Please help. Any suggestion is appreciated!!!

Thanks in advance.

<img width="761" height="629" alt="Image" src="https://github.com/user-attachments/assets/0f99fb3a-67ee-4b4a-8ae2-58f8a1af0732" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @dilwolf, thank you for reaching out and providing detailed context and code snippets! This is an automated response to help get you the fastest support possible. An Ultralytics engineer will review and assist you further soon. 🚀

We recommend checking the [Docs](https://docs.ultralytics.com/) for information on ONNX exports, OBB tasks, and preprocessing, as well as usage examples for both [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if you haven’t already. This helps us debug and compare the results precisely.

If you have questions about custom training or inference ❓, please provide as much information as possible, including:
- Example images and expected vs. actual results
- Your full environment details (Python, PyTorch, ONNX Runtime versions)
- Any logs or error messages

Join the Ultralytics community for more help or discussion:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth Q&A: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to confirm the issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience! An Ultralytics engineer will follow up to help diagnose your OBB ONNX output differences.

**Comment by Y-T-G** [comment]
You can find the correct pre-processing code here: 

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-ONNXRuntime

**Comment by dilwolf** [comment]
Thank you so much. It helped me. I found out that I was not doing letterbox properly

**Comment by glenn-jocher** [comment]
Great to hear you solved it! Letterbox preprocessing is indeed crucial for maintaining aspect ratios and ensuring consistent results between custom implementations and Ultralytics inference. The examples in the ONNXRuntime directory show the complete preprocessing pipeline including proper letterbox resizing which is essential for accurate model outputs.
